### DISCUSSION TITLE
How to create and link two records in one transaction.

### DISCUSSION THREAD

**Discussion by Izhaki** [planetscale/discussion#385] [Q&A] [answered]
Both this [answer](https://github.com/planetscale/discussion/discussions/252) and the [linked docs](https://planetscale.com/docs/reference/mysql-compatibility) assert that procedures and functions are not supported.

# Use Case

Consider we model polls (voting). The user creates a pool, which can have a few answers (note that users vote for answers, not a poll).

<img src="https://user-images.githubusercontent.com/880132/216787950-d5fa21b3-8a9f-4d0b-b9cc-4dd8df42e82d.png" width="300">

There are two tables involved:

- Polls
- Answers

We want:
- First create the poll
- Get its newly created ID
- Create each of the answers, linking to the poll ID

And all this needs to be a single transaction - if creating an answer fails, the poll insertion should be rolled back.

(This is just one of many examples that can be given.)

So, other than doing this at application level, is there a better way of doing this?

**Comment by CanRau** [comment]
Good question I'm curious as well as I'm currently playing around with pscale, converting my psql schema to evaluate it for our startup.
So far (PostgreSQL) I am using in most cases nanoid as primary key which I generate beforehand in code and can then create the main record and then link it later in other records. From Planetscale [blog](https://planetscale.com/blog/why-we-chose-nanoids-for-planetscales-api), I think also in some issue or discussion here on GH it seems auto incrementing bigint is recommended for better performance and/at scale, so if I'd now follow this recommendation what's the best approach, other then creating, then selecting and then linking?

Also stumbled upon this unanswered SO [Does Planetscale have LAST_INSERT_ID functionality?](https://stackoverflow.com/questions/73200864/does-planetscale-have-last-insert-id-functionality)

**Edit**: One alternative might be something like ksuid or ulid, cause I really dislike uuid, visually and for the reasons mentioned in the linked pscale blog post

**Reply by mscoutermarsh** [reply]
I'd highly recommend using an auto incrementing big int. Then you can use `last_insert_id()`. To grab it. Check an example here: https://github.com/planetscale/discussion/discussions/385#discussioncomment-4918806

As far as nanoid's etc. We use them as well, but only as a public facing ID.

**Comment by CanRau** [comment]
Isn't it also useful for sharding to use sthg like nanoid, ksuid or similar, to ensure uniqueness across shards, at least that's what I'm reading here & there or would it be sufficient to use a composite of shard key + row id?

**Reply by mscoutermarsh** [reply]
Great question. I generally wouldn't recommend starting a new app with plans to shard unless you expect a lot of data. Needing sharding is a good problem to have of course. But it's not worth being too worried about until it's more of a problem for you.

Having a general understanding of how it works though can help in setting up your app to work well in the future.

The way to figure out your shard key is to ask "how is ownership of the data grouped?". 

The simple answer often is, whatever "owns" the data, is a great shard key.

**Example**
GitHub.com is a good example. All data on github is "owned" by the `organization`. Even this webpage, you can see the `planetscale` organization name is in the URL.

In their database, the org has an id. Such as `5`. For sharding, you'd then add `organization_id` to all of your sharded tables. Then when querying a sharded table, you'd always include `organization_id = 5`. And Vitess would know which shard to look at to grab the correct data.

So if you're wanting to future proof an app. The simple step to take is to have your URLs setup like GitHub's. Whatever entity "owns" the data, being in the URL, will give every web request easy access to the shard key.

**Reply by CanRau** [reply]
Yea I agree, was just thinking as at least some resources seem to suggest unique identifiers or per table calculated auto-incrementing sequences like Instagram that it might be better to not use "simple" auto-increments, then I thought why not use unique ids from the get go, that's why I was curious what your recommendation then is, thanks a lot for explaining 🙏🏼 

How would you map `5` to `planetscale` though ?

**Reply by mscoutermarsh** [reply]
> How would you map 5 to planetscale though?

The organizations table could be unsharded. Unlikely that someone would have terabytes of org data.

`select id from organizations where slug = 'planetscale'`. Then use the ID from there.

**Reply by CanRau** [reply]
Uh yea I see, makes sense 👌🏼

**Reply by CanRau** [reply]
Got a follow up thought on this, even though sharding is probably still far ahead in the future for us. In your example adding a sharding key which would also be an integer, you'd then use the combination of former pk and sharding key as new pk, or how would you avoid collisions in shards?
Just came across Cuid2's [Note on K-Sortable/Sequential/Monotonically Increasing Ids](https://github.com/paralleldrive/cuid2#note-on-k-sortablesequentialmonotonically-increasing-ids) and got curious again 😅 
Also came across [this piece](https://dev.to/yugabyte/uuid-or-cached-sequences-42fi) from Yugabyte which is also in favor of sequences

**Reply by mscoutermarsh** [reply]
When it's time to shard, you'd replace auto increment with a Vitess sequence (vitess tracks the ID for you in a separate table)

**Reply by CanRau** [reply]
Aaah fantastic didn't know about this, thanks a lot 🙏

We're currently preparing the migration 🤓

**Comment by mscoutermarsh** [comment, marked as answer]
Hi @Izhaki,

I'd recommend looking at [transactions](https://www.mysqltutorial.org/mysql-transaction.aspx) and [last_insert_id](https://www.mysqltutorial.org/mysql-last_insert_id.aspx).

If you're writing this in direct SQL. It could work something like this. 

```sql
START TRANSACTION;
insert into polls (question) values ("What is my favorite color?");
insert into answers (poll_id, answer_text) values (last_insert_id(), "red"), (last_insert_id(), "blue");
COMMIT;
```

Each of these lines you'd submit separately on the same connection. If one of the inserts failed, you'd call `ROLLBACK;`

(I just wrote this out without testing, so it may have bugs, but I hope you get the general idea).

Generally, I'd use an ORM for this instead to make it much easier. If I were doing this in Rails it would look like.

```ruby
user.transaction do
   poll = user.polls.create!(question: "what is my favorite color?")
   poll.answers.create!(answer_text: "red")
   poll.answers.create!(answer_text: "blue")
end
```

Rails would be submitting essentially the same SQL I showed above, but would auto handle errors and issuing the rollback.

Hope that helps!

**Reply by Izhaki** [reply]
Thanks. This works.

Should probably also mention the transaction API for the database-js driver: https://github.com/planetscale/database-js#transactions
